src/k8s_upgrade.py

```diff
@@ -0,0 +1,487 @@
+
+        if self._upgrade.unit_state is UnitState.RESTARTING:  # Kubernetes only
+            if not self._upgrade.is_compatible:
+                logger.info(
+                    "Upgrade incompatible. If you accept potential *data loss* and *downtime*, you can continue with `resume-upgrade force=true`"
+                )
+                self.charm.status.set_and_share_status(Config.Status.UNHEALTHY_UPGRADE)
+                return
+        if self.charm.db_initialised and self.charm.is_db_service_ready():
+            self._upgrade.unit_state = UnitState.HEALTHY
+        if self.charm.unit.is_leader():
+            self._upgrade.reconcile_partition()
+
+        self._set_upgrade_status()
+
+    def _set_upgrade_status(self):
+        if self.charm.unit.is_leader():
+            self.charm.app.status = self._upgrade.app_status or ActiveStatus()
+        # Set/clear upgrade unit status if no other unit status - upgrade status for units should
+        # have the lowest priority.
+        if isinstance(self.charm.unit.status, ActiveStatus) or (
+            isinstance(self.charm.unit.status, BlockedStatus)
+            and self.charm.unit.status.message.startswith(
+                "Rollback with `juju refresh`. Pre-upgrade check failed:"
+            )
+        ):
+            self.charm.status.set_and_share_status(
+                self._upgrade.get_unit_juju_status() or ActiveStatus()
+            )
+
+    def _on_upgrade_peer_relation_created(self, _) -> None:
+        """First time the relation is created, we save the revisions."""
+        if self.charm.unit.is_leader():
+            self._upgrade.set_versions_in_app_databag()
+
+    def _on_resume_upgrade_action(self, event: ActionEvent) -> None:
+        if not self.charm.unit.is_leader():
+            message = f"Must run action on leader unit. (e.g. `juju run {self.charm.app.name}/leader {RESUME_ACTION_NAME}`)"
+            logger.debug(f"Resume upgrade event failed: {message}")
+            event.fail(message)
+            return
+        if not self._upgrade or not self._upgrade.in_progress:
+            message = "No upgrade in progress"
+            logger.debug(f"Resume upgrade event failed: {message}")
+            event.fail(message)
+            return
+        self._upgrade.reconcile_partition(action_event=event)
+
+    def run_post_app_upgrade_task(self, event: EventBase):
+        """Runs the post upgrade check to verify that the cluster is healthy.
+
+        By deferring before setting unit state to HEALTHY, the user will either:
+            1. have to wait for the unit to resolve itself.
+            2. have to run the force-upgrade action (to upgrade the next unit).
+        """
+        logger.debug("Running post upgrade checks to verify cluster is not broken after upgrade")
+        self.run_post_upgrade_checks(event, finished_whole_cluster=False)
+
+        if self._upgrade.unit_state != UnitState.HEALTHY:
+            return
```

Comment:
noob q: why do we return (i.e. not defer here) perhaps we could log / add a comment here

Reply:
This is just a copy/paste from the mongodb vm upgrades, we're not continuing the overall cluster checks if we're not healthy in our local replicaset.

Reply:
Added a comment.